File: R/DRYAIR.R
#' DRYAIR
#'
#' Calculates several properties of dry air and related characteristics shownas output variables below. The program
#' is based on equations from List, R. J. 1971. Smithsonian Meteorological Tables. Smithsonian
#' Institution Press. Washington, DC. WETAIR must be used in conjunction with function VAPPRS.
#'
#' The user must supply values for the input variables (db, bp and alt).
#' If alt is known (-1000 < alt < 20000) but not BP, then set BP=0
#' @param db Dry bulb temperature (degrees C)
#' @param bp Barometric pressure (pascal)
#' @param alt Altitude (m)
#' @return patmos Standard atmospheric pressure (Pa)
#' @return densty Density (kg m-3)
#' @return visdyn Dynamic viscosity (kg m-1 s-1)
#' @return viskin Kinematic viscosity (m2 s-1)
#' @return difvpr Diffusivity of water vapour in air (m2 s-1)
#' @return thcond Thermal conductivity (W m-1 K-1)
#' @return htovpr Latent heat of vapourisation of water (J kg-1)
#' @return tcoeff Temperature coefficient of volume expansion (K-1)
#' @return ggroup Group of variables in Grashof number (1-m3 -K)
#' @return bbemit black body emittance (W m-2)
#' @return emtmax Wave length of maximum emittance (m)
#' @export
DRYAIR <- function(db=db, bp=0, alt=0){
  tstd=273.15
  pstd=101325.
  patmos=pstd*((1-(0.0065*alt/288))^(1/0.190284))
  bp=rep(bp,length(patmos))
  bp[bp<=0]=patmos[bp<=0]
  densty=bp/(287.04*(db+tstd))
  visnot=1.8325E-5
  tnot=296.16
  c=120.
  visdyn=(visnot*((tnot+c)/(db+tstd+c)))*(((db+tstd)/tnot)^1.5)
  viskin=visdyn/densty
  difvpr=2.26E-5*(((db+tstd)/tstd)^1.81)*(1.E5/bp)
  thcond=0.02425+(7.038E-5*db)
  htovpr=2.5012E6-2.3787E3*db
  tcoeff=1./(db+tstd)
  ggroup=0.0980616*tcoeff/(viskin*viskin)
  bbemit=5.670367E-8*((db+tstd)^4)
  emtmax=2.897E-3/(db+tstd)
  return(list(patmos=patmos, densty=densty, visdyn=visdyn, viskin=viskin, difvpr=difvpr,
    thcond=thcond, htovpr=htovpr, tcoeff=tcoeff, ggroup=ggroup, bbemit=bbemit, emtmax=emtmax))
}

File: src/DRYAIR.f
      SUBROUTINE DRYAIR(DB,BP,ALT,PATMOS,DENSTY,VISDYN,VISKIN,DIFVPR,   
     *THCOND,HTOVPR,TCOEFF,GGROUP)   
     
C     NicheMapR: software for biophysical mechanistic niche modelling

C     Copyright (C) 2018 Michael R. Kearney and Warren P. Porter

c     This program is free software: you can redistribute it and/or modify 
c     it under the terms of the GNU General Public License as published by 
c     the Free Software Foundation, either version 3 of the License, or (at
c      your option) any later version.

c     This program is distributed in the hope that it will be useful, but
c     WITHOUT ANY WARRANTY; without even the implied warranty of 
c     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU 
c     General Public License for more details.

c     You should have received a copy of the GNU General Public License 
c     along with this program. If not, see http://www.gnu.org/licenses/.

c     Subroutine wetair calculates several properties of humid air.  
c     This version was taken from 
c     Tracy, C. R., W. R. Welch, B. Pinshow, M. R. Kearney, and W. P. 
c     Porter. 2016. Properties of air: A manual for use in biophysical ecology. 
c     5th edition. The University of Wisconsin, Madison. 
c     which is available as a vignette in the NicheMapR package
     
C***********************************************************************
C   
C SUBROUTINE DRYAIR CALCULATES SEVERAL PROPERTIES OF DRY AIR AND RELATED
C CHARACTERISTICS SHOWN AS OUTPUT VARIABLES BELOW.  THE PROGRAM IS BASED
C ON DATA FROM LIST, R. J. 1971. SMITHSONIAN METEOROLOGICAL TABLES. 
C SMITHSONIAN INSTITUTION PRESS. WASHINGTON, DC.
C   
C THE USER MUST SUPPLY VALUES FOR THE INPUT VARIABLES (DB, BP, AND ALT).
C IF ALT IS KNOWN (-1 000 < ALT < 20 000) BUT NOT BP THEN SET BP=0. 
C   
C*************************** INPUT VARIABLES ***************************
C   
C DB=DRY BULB TEMPERATURE (DEGREE CELSIUS)  
C BP=BAROMETRIC PRESSURE (PASCAL) [BP AT ONE STANDARD ATMOSPHERE IS 
C    101 325 PASCALS (100 PASCALS=1 MILLIBAR)]. 
C ALT=ALTITUDE (METRE) (1 METRE=3.280 839 9 FEET)   
C   
C*************************** OUTPUT VARIABLES **************************
C   
C PATMOS=STANDARD ATMOSPHERIC PRESSURE (PASCAL) 
C DENSTY=DENSITY (KILOGRAM PER CUBIC METRE) 
C VISDYN=DYNAMIC VISCOSITY (KILOGRAM PER METRE SECOND)  
C VISKIN=KINEMATIC VISCOSITY (SQUARE METRE PER SECOND)  
C DIFVPR=DIFFUSIVITY OF WATER VAPOR IN AIR (SQUARE METRE PER SECOND)
C THCOND=THERMAL CONDUCTIVITY (WATT PER METRE KELVIN)   
C HTOVPR=LATENT HEAT OF VAPORIZATION OF WATER (JOULE PER KILOGRAM)  
C TCOEFF=TEMPERATURE COEFFICIENT OF VOLUME EXPANSION (1 PER KELVIN) 
C GGROUP=GROUP OF VARIABLES IN GRASHOF NUMBER (1 PER CUBIC METRE KELVIN)
C   
C***********************************************************************
C   
      IMPLICIT LOGICAL (A-Z)
      DOUBLE PRECISION ALT,BP,C,DB,DENSTY,DIFVPR,GGROUP,HTOVPR,PATMOS
     *,PSTD,TCOEFF,THCOND,TNOT,TSTD,VISDYN,VISKIN,VISNOT  

      TSTD=273.15   
      PSTD=101325.  
      PATMOS=PSTD*((1.-(.0065*ALT/288.))**(1./.190284)) 
      IF(BP .LE. 0.)BP=PATMOS   
      DENSTY=BP/(287.04*(DB+TSTD))  
      VISNOT=1.8325D-5  
      TNOT=296.16   
      C=120.
      VISDYN=(VISNOT*((TNOT+C)/(DB+TSTD+C)))*(((DB+TSTD)/TNOT)**1.5)
      VISKIN=VISDYN/DENSTY  
      DIFVPR=2.26D-5*(((DB+TSTD)/TSTD)**1.81)*(1.D5/BP) 
      THCOND=.02425+(7.038D-5*DB)   
      HTOVPR=2.5012D6-2.3787D3*DB   
      TCOEFF=1./(DB+TSTD)   
      GGROUP=.0980616*TCOEFF/(VISKIN*VISKIN)
      RETURN
      END
